server/src/main/java/org/elasticsearch/script/Metadata.java

```diff
@@ -50,80 +46,31 @@ public class Metadata {
     protected static final String IF_PRIMARY_TERM = "_if_primary_term";
     protected static final String DYNAMIC_TEMPLATES = "_dynamic_templates";
 
-    protected static final Map<String, Validator> VALIDATORS = Map.of(
-        INDEX,
-        Metadata::stringValidator,
-        ID,
-        Metadata::stringValidator,
-        ROUTING,
-        Metadata::stringValidator,
-        VERSION_TYPE,
-        Metadata::versionTypeValidator,
-        VERSION,
-        Metadata::notNullLongValidator,
-        TYPE,
-        Metadata::stringValidator,
-        IF_SEQ_NO,
-        Metadata::longValidator,
-        IF_PRIMARY_TERM,
-        Metadata::longValidator,
-        DYNAMIC_TEMPLATES,
-        Metadata::mapValidator
-    );
-
     protected final Map<String, Object> map;
-    protected final Map<String, Validator> validators;
-
-    // timestamp is new to ingest metadata, so it doesn't need to be backed by the map for back compat
-    protected final ZonedDateTime timestamp;
-
-    public Metadata(String index, String id, long version, String routing, VersionType versionType, ZonedDateTime timestamp) {
-        this(metadataMap(index, id, version, routing, versionType), timestamp, VALIDATORS);
-    }
+    protected final Map<String, FieldProperty<?>> properties;
+    protected static final FieldProperty<?> BAD_KEY = new FieldProperty<>(null, false, false, null);
 
-    public Metadata(Map<String, Object> map, ZonedDateTime timestamp) {
-        this(map, timestamp, VALIDATORS);
-    }
-
-    Metadata(Map<String, Object> map, ZonedDateTime timestamp, Map<String, Validator> validators) {
+    public Metadata(Map<String, Object> map, Map<String, FieldProperty<?>> properties) {
         this.map = map;
-        this.timestamp = timestamp;
-        this.validators = validators;
+        this.properties = Collections.unmodifiableMap(properties);
         validateMetadata();
     }
 
-    /**
-     * Create the backing metadata map with the standard contents assuming default validators.
-     */
-    protected static Map<String, Object> metadataMap(String index, String id, long version, String routing, VersionType versionType) {
-        Map<String, Object> metadata = Maps.newHashMapWithExpectedSize(IngestDocument.Metadata.values().length);
-        metadata.put(IngestDocument.Metadata.INDEX.getFieldName(), index);
-        metadata.put(IngestDocument.Metadata.ID.getFieldName(), id);
-        metadata.put(IngestDocument.Metadata.VERSION.getFieldName(), version);
-        if (routing != null) {
-            metadata.put(IngestDocument.Metadata.ROUTING.getFieldName(), routing);
-        }
-        if (versionType != null) {
-            metadata.put(IngestDocument.Metadata.VERSION_TYPE.getFieldName(), VersionType.toString(versionType));
-        }
-        return metadata;
-    }
-
     /**
      * Check that all metadata map contains only valid metadata and no extraneous keys and source map contains no metadata
```

Comment:
nit: I think this comment is off? there is no source map

Reply:
Updated.